src/lib/schedule.c

```diff
@@ -25,459 +25,95 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  *
- * Author: Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ * Author: Bartosz Kokoszko <bartoszx.kokoszko@linux.intel.com>
  */
 
-#include <stdint.h>
-#include <stddef.h>
-#include <errno.h>
-#include <sof/sof.h>
-#include <sof/lock.h>
-#include <sof/list.h>
-#include <sof/stream.h>
-#include <sof/alloc.h>
-#include <sof/debug.h>
-#include <sof/clk.h>
+/* Generic scheduler */
 #include <sof/schedule.h>
-#include <sof/work.h>
-#include <platform/timer.h>
-#include <platform/clk.h>
-#include <sof/audio/component.h>
-#include <sof/drivers/timer.h>
-#include <sof/task.h>
+#include <sof/edf_schedule.h>
+#include <sof/ll_schedule.h>
 
-struct schedule_data {
-	spinlock_t lock;
-	struct list_item list;	/* list of tasks in priority queue */
-	uint32_t clock;
-	struct work work;
+struct scheduler_ops *schedulers[SOF_TASKS_COUNT] = {
+	&schedule_edf_ops,              /* SOF_EDF_TASK */
+	&schedule_ll_ops		/* SOF_LOW_LATENCY_TASK */
 };
 
-#define SLOT_ALIGN_TRIES	10
-
-/*
- * Simple rescheduler to calculate tasks new start time and deadline if
- * prevoius deadline was missed. Tries to align at first with current task
- * timing, but will just add onto current if too far behind current.
- * XRUNs will be propagated upto the host if we have to reschedule.
- */
-static inline void edf_reschedule(struct task *task, uint64_t current)
+void schedule_task_init(struct task *task, uint32_t type, uint32_t priority,
+			uint64_t (*func)(void *data), void *data,
+			uint32_t core, uint32_t xflags)
 {
-	uint64_t delta = (task->deadline - task->start) << 1;
-	int i;
-
-	/* try and align task with current scheduling slots */
-	for (i = 0; i < SLOT_ALIGN_TRIES; i++) {
-
-		task->start += delta;
-
-		if (task->start > current + delta) {
-			task->deadline = task->start + delta;
-			return;
-		}
-	}
-
-	/* task has slipped a lot, so just add delay to current */
-	task->start = current + delta;
-	task->deadline = task->start + delta;
+	task->type = type;
+	task->priority = priority;
+	task->core = core;
+	task->state = SOF_TASK_STATE_INIT;
+	task->func = func;
+	task->data = data;
+
+	schedulers[task->type]->schedule_task_init(task, xflags);
```

Comment:
type needs to be validated.

Reply:
in fact we need to have a task->ops (that maps to scheduler ops) set here.

Reply:
I've added ops to task structure. Task type is validated now in init function.